File: app/src/main/java/com/letianpai/robot/wificonnet/view/KeyButton.java
package com.letianpai.robot.wificonnet.view;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;

import com.letianpai.robot.wificonnet.R;

/**
 * @author liujunbin
 */
public class KeyButton extends Button {
    private KeyButton keyButton;
    private Context mContext;
    public KeyButton(Context context) {
        super(context);
        init(context);
    }

    private void init(Context context) {
        this.mContext = context;
        keyButton = KeyButton.this;
        setButtonUnPressed();
        addKeyButtonListeners();
    }

    public KeyButton(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public KeyButton(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    public KeyButton(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context);
    }


    private void addKeyButtonListeners() {

        keyButton.setOnTouchListener(new OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                if (event.getAction() == MotionEvent.ACTION_UP){
                    setButtonUnPressed();

                }else if (event.getAction() == MotionEvent.ACTION_CANCEL){
                    setButtonUnPressed();

                }else if (event.getAction() == MotionEvent.ACTION_DOWN){
                    setButtonPressed();
                }
                return false;
            }
        });
    }

    private void setButtonUnPressed() {
        keyButton.setBackgroundColor(mContext.getResources().getColor(R.color.background));
        keyButton.setTextColor(mContext.getColor(R.color.keyboard_text_color));
    }

    private void setButtonPressed() {
        keyButton.setBackgroundColor(mContext.getResources().getColor(R.color.keyboard_highlight));
        keyButton.setTextColor(mContext.getColor(R.color.white));
        
    }


}

File: app/src/main/java/com/letianpai/robot/wificonnet/view/KeyButton.kt
package com.letianpai.robot.wificonnet.view

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.widget.Button
import com.letianpai.robot.wificonnet.R

/**
 * @author liujunbin
 */
class KeyButton : androidx.appcompat.widget.AppCompatButton {
    private var keyButton: KeyButton? = null
    private var mContext: Context? = null

    constructor(context: Context) : super(context) {
        init(context)
    }

    private fun init(context: Context) {
        this.mContext = context
        keyButton = this@KeyButton
        setButtonUnPressed()
        addKeyButtonListeners()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init(context)
    }

    private fun addKeyButtonListeners() {
        keyButton!!.setOnTouchListener { v, event ->
            if (event.action == MotionEvent.ACTION_UP) {
                setButtonUnPressed()
            } else if (event.action == MotionEvent.ACTION_CANCEL) {
                setButtonUnPressed()
            } else if (event.action == MotionEvent.ACTION_DOWN) {
                setButtonPressed()
            }
            false
        }
    }

    private fun setButtonUnPressed() {
        keyButton!!.setBackgroundColor(mContext!!.resources.getColor(R.color.background))
        keyButton!!.setTextColor(mContext!!.getColor(R.color.keyboard_text_color))
    }

    private fun setButtonPressed() {
        keyButton!!.setBackgroundColor(mContext!!.resources.getColor(R.color.keyboard_highlight))
        keyButton!!.setTextColor(mContext!!.getColor(R.color.white))
    }
}
